### PR TITLE
ENH: Add proper dtype-support to `np.flatiter`

### PR DESCRIPTION
**Pull request by BvB93**
This PR adds proper dtype-support to `np.flatiter`.

The methods of the latter now successfully return an array/scalar of the appropriate dtype (when applicable).

Note that In order to get `__array__` to work, the `npt._SupportsArray` protocol needed some minor restructuring,
the latter now only caring about the default value of its `dtype` parameter (i.e. `dtype=None`).
Concrete implementations of the protocol are responsible for adding any and all `DTypeLike`-based overloads.

Examples
----------
Given the `np.int64` array `i8_array`:
* `next(i8_array.flat)` and `i8_array.flat[0]` are now inferred as `np.int64`.
* `i8_array.flat.__array__()` is now inferred as `np.ndarray[Any, np.dtype[np.int64]]`.

### CODE REVIEW

File: numpy/typing/_array_like.py [hunk 1]
```diff
@@ -16,12 +18,15 @@
     else:
         HAVE_PROTOCOL = True
 
+_DType = TypeVar("_DType", bound="dtype[Any]")
+
 if TYPE_CHECKING or HAVE_PROTOCOL:
-    class _SupportsArray(Protocol):
-        @overload
```
Comment (BvB93) [inline numpy/typing/_array_like.py:21]: Turns out the protocol already works fine for both positional-only and positional-or-keyword arguments, even without these overloads.

File: numpy/__init__.pyi [hunk 2]
```diff
@@ -955,24 +955,30 @@ _ArrayLikeInt = Union[
...
     @overload
     def __getitem__(
         self, key: Union[_ArrayLikeInt, slice, ellipsis],
-    ) -> _ArraySelf: ...
-    def __array__(self, __dtype: DTypeLike = ...) -> ndarray: ...
+    ) -> _NdArraySubClass: ...
+    @overload
+    def __array__(self: flatiter[ndarray[Any, _DType]], __dtype: None = ...) -> ndarray[Any, _DType]: ...
+    @overload
+    def __array__(self, __dtype: DTypeLike) -> ndarray[Any, dtype[Any]]: ...
```
Comment (BvB93) [inline numpy/__init__.pyi:981]: Adding overloads for all `DTypeLike` objects is quite frankly too much work. 
For now just return `dtype[Any]`  if a dtype-like object (besides `None`) is provided.

File: numpy/typing/_array_like.py [hunk 3]
```diff
@@ -16,12 +18,15 @@
...
-        def __array__(self, __dtype: DTypeLike = ...) -> ndarray: ...
-        @overload
-        def __array__(self, dtype: DTypeLike = ...) -> ndarray: ...
+    # The `_SupportsArray` protocol only cares about the default dtype
+    # (i.e. `dtype=None`) ofthe to-be returned array.
```
Comment (charris) [inline numpy/typing/_array_like.py:25]: typo "ofthe".

Comment (BvB93) [reply numpy/typing/_array_like.py]: Good call, fixed as of https://github.com/numpy/numpy/pull/17981/commits/6139ed42af271d37234320e83a41000d93bdeae1.

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**